Add sending allowed children notification article (#4487)
umbraco/UmbracoDocs: Reference/Notifications/SendingAllowedChildrenNotification/index.md

---
versionFrom: 9.5.0
versionTo: 10.0.0
---

# Sending Allowed Children Notification

> The `SendingAllowedChildrenNotification` enable you to manipulate the content types that will be shown in the create menu.

Reviewer: ```suggestion

The `SendingAllowedChildrenNotification` enables you to manipulate the content types that will be shown in the create menu when adding new content in the backoffice.
```

## Usage

> Example usage of the `SendingAllowedChildrenNotification` - for example show only the settings content type if none have been added yet.

Reviewer: How about rephrasing this part to:

```suggestion

With the example below we can ensure that a content type cannot be selected if the type already exists in the Content tree.
```

    {
        public void Handle(SendingAllowedChildrenNotification notification)
        {
            const string contentIdKey = "contentId";

            // Try get the id from the content item in the backoffice 
            var queryStringCollection = HttpUtility.ParseQueryString(notification.UmbracoContext.OriginalRequestUrl.Query);

            if (!queryStringCollection.ContainsKey(contentIdKey))
            {
                return;
            }

            var contentId = queryStringCollection[contentIdKey].TryConvertTo<int>().ResultOr(-1);

            if (contentId == -1)
            {
                return;
            }

            var content = notification.UmbracoContext.Content?.GetById(true, contentId);

            if (content is null)
            {
                return;
            }

            // Allowed children as configured in the backoffice
            var allowedChildren = notification.Children.ToList();

            if (content.ChildrenForAllCultures is not null)
            {
                // Get all children of current page
                var childNodes = content.ChildrenForAllCultures.ToList();

>                 // If there is a Settings page already created, then don't allow it anymore

Reviewer 2: In case you want to show the `Settings.ModelTypeAlias` example you had:

```suggestion
                // If there is a Settings page already created, then don't allow it anymore
                // You can also use the ModelTypeAlias property from your PublishedModel for comparison,
                // like Settings.ModelTypeAlias if you have set models builder to generate SourceCode models
```

                if (childNodes.Any(x => x.ContentType.Alias == "settings"))
                {
                    allowedChildren.RemoveAll(x => x.Alias == "settings");
                }
            }

            // Update the allowed children
            notification.Children = allowedChildren;
        }
    }
}

> ```

Reviewer 2: After the code example you can mention:

> You also need to register this notification handler, like:
> 
> ```csharp
> .AddNotificationHandler<SendingAllowedChildrenNotification, SendingAllowedChildrenNotificationHandler>()
>                 
> ```

            {
                return;
            }

            // Allowed children as configured in the backoffice
            var allowedChildren = notification.Children.ToList();

            if (content.ChildrenForAllCultures is not null)
            {
                // Get all children of current page
                var childNodes = content.ChildrenForAllCultures.ToList();

                // If there is a Settings page already created, then don't allow it anymore
                // You can also use the ModelTypeAlias property from your PublishedModel for comparison,
                // like Settings.ModelTypeAlias if you have set models builder to generate SourceCode models
                if (childNodes.Any(x => x.ContentType.Alias == "settings"))
                {
                    allowedChildren.RemoveAll(x => x.Alias == "settings");
                }
            }

            // Update the allowed children
            notification.Children = allowedChildren;
        }
    }
}
```

You also need to register this notification handler. You can achieve this by updating the `Startup` class like:

```csharp
public void ConfigureServices(IServiceCollection services)

> {
> #pragma warning disable IDE0022 // Use expression body for methods
>     services.AddUmbraco(_env, _config)
>         .AddBackOffice()             
>         .AddWebsite()
>         .AddComposers()
>         .AddNotificationHandler<SendingAllowedChildrenNotification, SendingAllowedChildrenNotificationHandler>()
>         .Build();
> #pragma warning restore IDE0022 // Use expression body for methods
> }

Reviewer 2: ```suggestion
{
    services.AddUmbraco(_env, _config)
        .AddBackOffice()             
        .AddWebsite()
        .AddComposers()
        .AddNotificationHandler<SendingAllowedChildrenNotification, SendingAllowedChildrenNotificationHandler>()
        .Build();
}
```

```

For more information av

:::note
The license for Umbraco Deploy comes with a recurring yearly fee. Learn more about this and pricing on [Umbraco.com](https://umbraco.com/products/umbraco-deploy/).

> :::

Reviewer 2: I believe this is a wrong paste and perhaps you want to refer to https://our.umbraco.com/documentation/Reference/Notifications/#registering-notification-handlers instead 😉

Author: oopss... 🙈 my bad. Just updated 😄
